sdk/core/Azure.Core.TestFramework/src/OperationInterceptor.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using Castle.DynamicProxy;
+
+namespace Azure.Core.TestFramework
+{
+    internal class OperationInterceptor : IInterceptor
+    {
+        private static readonly object NoWaitDelay = TimeSpan.Zero;
+        private static readonly string WaitForCompletionMethodName = nameof(Operation<object>.WaitForCompletionAsync);
+        private static readonly MethodInfo WaitForCompletionResponseAsync = typeof(Operation).GetMethod(nameof(Operation.WaitForCompletionResponseAsync), new[]{typeof(TimeSpan), typeof(CancellationToken)});
+
+        private readonly bool _noWait;
+
+        public OperationInterceptor(bool noWait)
+        {
+            _noWait = noWait;
+        }
+
+        public void Intercept(IInvocation invocation)
+        {
+            if (_noWait)
+            {
+                if (invocation.Method.Name == WaitForCompletionMethodName)
+                {
+                    CheckArguments(invocation.Arguments);
+                    var cancellationToken = invocation.Arguments.Last();
+                    var waitForCompletionMethod = invocation.TargetType.GetMethod(WaitForCompletionMethodName, new[]{typeof(TimeSpan), typeof(CancellationToken)});
```

Comment:
Why do we store the `WaitForCompletionResponseAsync` and not the sync MethodInfo?

Reply:
WaitForCompletionMethodName is on generic type so we have to get closed generic version of it for every call.